api/src/test/java/org/pmiops/workbench/workspaces/WorkspaceAuthServiceTest.java

```diff
@@ -0,0 +1,288 @@
+
+  @Test
+  public void test_validateActiveBilling_active() {
+    final String namespace = "wsns";
+    final String fcName = "firecloudname";
+    stubDaoGetRequired(namespace, fcName, BillingStatus.ACTIVE);
+
+    // does not throw
+    workspaceAuthService.validateActiveBilling(namespace, fcName);
+  }
+
+  @Test
+  public void test_validateActiveBilling_inactive() {
+    final String namespace = "wsns";
+    final String fcName = "firecloudname";
+    stubDaoGetRequired(namespace, fcName, BillingStatus.INACTIVE);
+
+    assertThrows(
+        ForbiddenException.class,
+        () -> workspaceAuthService.validateActiveBilling(namespace, fcName));
+  }
+
+  private static Stream<Arguments> accessLevels() {
+    return Stream.of(
+        Arguments.of("OWNER", WorkspaceAccessLevel.OWNER),
+        Arguments.of("WRITER", WorkspaceAccessLevel.WRITER),
+        Arguments.of("READER", WorkspaceAccessLevel.READER),
+        Arguments.of("NO ACCESS", WorkspaceAccessLevel.NO_ACCESS),
+        Arguments.of("PROJECT_OWNER", WorkspaceAccessLevel.OWNER));
+  }
+
+  @ParameterizedTest(name = "getWorkspaceAccessLevel({0})")
+  @MethodSource("accessLevels")
+  public void test_getWorkspaceAccessLevel_valid(
+      String accessLevel, WorkspaceAccessLevel expected) {
+    final String namespace = "wsns";
+    final String fcName = "firecloudname";
+    stubFcGetWorkspace(namespace, fcName, accessLevel);
+    assertThat(workspaceAuthService.getWorkspaceAccessLevel(namespace, fcName)).isEqualTo(expected);
+  }
+
+  @Test
+  public void test_getWorkspaceAccessLevel_invalid() {
+    final String namespace = "wsns";
+    final String fcName = "firecloudname";
+    stubFcGetWorkspace(namespace, fcName, "some garbage");
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> workspaceAuthService.getWorkspaceAccessLevel(namespace, fcName));
+  }
+
+  private static Stream<Arguments> enforcedAccessLevels_valid() {
+    return Stream.of(
+        Arguments.of("OWNER", WorkspaceAccessLevel.OWNER, WorkspaceAccessLevel.OWNER),
+        Arguments.of("OWNER", WorkspaceAccessLevel.OWNER, WorkspaceAccessLevel.READER),
+        Arguments.of("WRITER", WorkspaceAccessLevel.WRITER, WorkspaceAccessLevel.READER),
+        Arguments.of("READER", WorkspaceAccessLevel.READER, WorkspaceAccessLevel.READER),
+        Arguments.of("PROJECT_OWNER", WorkspaceAccessLevel.OWNER, WorkspaceAccessLevel.WRITER));
+  }
+
+  private static Stream<Arguments> enforcedAccessLevels_invalid() {
+    return Stream.of(
+        Arguments.of("WRITER", WorkspaceAccessLevel.OWNER, ForbiddenException.class),
+        Arguments.of("READER", WorkspaceAccessLevel.WRITER, ForbiddenException.class),
+        Arguments.of("NO ACCESS", WorkspaceAccessLevel.READER, ForbiddenException.class),
+        Arguments.of("NO ACCESS", WorkspaceAccessLevel.OWNER, ForbiddenException.class),
+        Arguments.of("invalid status", WorkspaceAccessLevel.READER, ServerErrorException.class));
+  }
```

Comment:
I found it's hard to read/add tests when some variables/methods breaks those test methods, shall we move all those to the top of the class as a static variable. And if there is any private method, suggest moving to bottom of the class, after all tests

Reply:
I agree for general helper functions, but for parameterized tests I want to make the association very clear and group these with the executable test methods